### PR TITLE
handle hard links and whiteouts correctly

### PR DESCRIPTION
**Pull request by alban**
Hard links and whiteout were not handled correctly in this scenario:
- The first Docker layer contains:
  - file A
  - file B hard link to file A
- The second Docker layer contains:
  - whiteout file A

Then, the squashed ACI previously generated by docker2aci contained:
  - file B dangling hard link to file A

Hence the bug.

This patch changes:
- the converting algorithm lib/common/common.go:writeACI() to generate:
  - First docker layer
    - .hidden.docker2aci.sha512-xxx (hash of the layer and file name)
    - file A hard link to .hidden.docker2aci.sha512-xxx
    - file B hard link to .hidden.docker2aci.sha512-xxx
  - Second docker layer
    - whiteout file A
- the squashing algorithm lib/docker2aci.go:SquashLayers() to have two
  passes:
  - Pass one: build an in-memory map of hard links and whiteouts
  - Pass two: remove white-out and .hidden.docker2aci.sha512-xxx files

I tested the following images from https://github.com/coreos/rkt/issues/1653:
- docker://albanc/busybox-hardlinks
- docker://zopyx/xmldirector-plone

TODO:
- not optimized if there are no hard links
- semantic changes on --nosquash

Fixes https://github.com/appc/docker2aci/issues/98

### CODE REVIEW

**Comment by alban** [comment]
@iaguis I think this is ready for a first review.

File: lib/docker2aci.go [hunk 1]
```diff
@@ -224,12 +224,42 @@ func getManifests(renderedACI acirenderer.RenderedACI, aciRegistry acirenderer.A
 	return manifests, nil
 }
 
-func writeSquashedImage(outputFile *os.File, renderedACI acirenderer.RenderedACI, aciProvider acirenderer.ACIProvider, manifests []schema.ImageManifest) error {
+func writeSquashedImage(outputFile *os.File,
+	renderedACI acirenderer.RenderedACI,
+	aciProvider acirenderer.ACIProvider,
+	manifests []schema.ImageManifest) error {
```
Comment (iaguis) [inline lib/docker2aci.go]: You don't touch the propotype anymore so you can leave the line as it was

**Comment by iaguis** [comment]
* Right now it only handles hardlinks to regular files, although weird, you could have a hardlink to a symlink
* I'm not concerned about this not being optimal for the no-hardlink case, but I am about the semantic changes in `--nosquash`. You mention that as a TODO, do you have any plans for it?
 
 Actually, it doesn't work. I just tested converting a busybox image with `--nosquash` and trying to run it in rkt results in

 ```
run: error setting up stage0: error setting up image \
sha512-2dffd725e98855c4afd53bbe2f053f7f3e94565fda1c7f6b753729998a997f47: \
error rendering tree image: treestore: cannot render aci: error extracting ACI: \
extracttar error: exit status 1, output: error extracting tar: error extracting tarball: link \
/rootfs/.hidden.docker2aci.sha512-fa4bf224598acc71239bc5ae543e9fa38f6f9032d2d219be2ba79f7a54259f505dcddc1b2ce3d8b643c8e83a356b24e837ced7637dc093d92bc88792af9c1947 \
/rootfs/bin/[: no such file or directory
 ```
 That happens because `.hidden.docker2aci.sha512-...` is not in the path white list so it doesn't get extracted and the hardlinks pointing to it are dangling.
 I don't think we should merge it in the current state because of that.

**Comment by alban** [comment]
Closing in favor of #105

**Pull request closed without merging**